server/src/main/java/com/exacaster/lighter/application/sessions/SessionHandler.java

```diff
@@ -51,28 +56,71 @@ public Waitable launch(Application application, Consumer<Throwable> errorHandler
-            var session = sessionService.fetchOne(sessionConf.getId());
+
+        final var allPermanentSessions = getPermanentSessionToCheck();
+
+        for (var perm : allPermanentSessions) {
+            var session = sessionService.fetchOne(perm.getSessionId());
             if (session.map(Application::getState).filter(this::running).isEmpty() ||
                     session.flatMap(backend::getInfo).map(ApplicationInfo::getState).filter(this::running).isEmpty()) {
-                LOG.info("Permanent session {} needs to be (re)started.", sessionConf.getId());
-                var sessionToLaunch = sessionService.createSession(
-                        sessionConf.getSubmitParams(),
-                        sessionConf.getId()
+                LOG.info("Permanent session {} needs to be (re)started.", perm.getSessionId());
+                sessionService.deletePermanentSession(perm.sessionId);
+                var sessionToLaunch = sessionService.createPermanentSession(
+                        perm.getSessionId(),
+                        perm.getSubmitParams()
                 );
 
-                sessionService.deleteOne(sessionToLaunch);
                 launchSession(sessionToLaunch).waitCompletion();
-                LOG.info("Permanent session {} (re)started.", sessionConf.getId());
+                LOG.info("Permanent session {} (re)started.", perm.getSessionId());
             }
         }
         LOG.info("End provisioning permanent sessions.");
     }
 
+    private List<PermanentSessionParam> getPermanentSessionToCheck() {
+        final var dbPermanentSessions = sessionService.fetchAllPermanentSessions();
+
+        final var configurationPermanentSessions = appConfiguration.getSessionConfiguration().getPermanentSessions().stream().collect(
+                Collectors.toMap(permanentSession -> permanentSession.getId(), Function.identity()));
+
+        final var fromYamlOnly = Sets.difference(configurationPermanentSessions.keySet(), dbPermanentSessions.keySet()).stream().map(
+                id -> new PermanentSessionParam(id, configurationPermanentSessions.get(id).getSubmitParams())
+        );
+
+        final var intersection = Sets.intersection(configurationPermanentSessions.keySet(), dbPermanentSessions.keySet()).stream()
+                .filter(id -> dbPermanentSessions.get(id).isNotDeleted())
+                .map(id -> new PermanentSessionParam(id, dbPermanentSessions.get(id).getSubmitParams()));
+
+        final var fromStorageOnly = Sets.difference(dbPermanentSessions.keySet(), configurationPermanentSessions.keySet()).stream().map(
+                id -> new PermanentSessionParam(id, dbPermanentSessions.get(id).getSubmitParams())
+        );
+
+        return Stream.concat(fromStorageOnly, Stream.concat(fromYamlOnly, intersection)).collect(Collectors.toList());
+
+    }
+
+    private static class PermanentSessionParam {
```

Comment:
Minor: Maybe it would be nicer for readability to move class definition to the bottom of the file, after all method definitions?

Reply:
done.